cmd/syncIndexer.go

```diff
@@ -21,17 +21,100 @@
 package cmd
 
 import (
+	"path/filepath"
 	"strings"
+	"sync"
+	"unsafe"
+
+	"github.com/nitin-deamon/azure-storage-azcopy/v10/common"
 )
 
+type folderIndexer struct {
+	folderMap map[string]*objectIndexer
+	lock      sync.Mutex
+	counter   int64
+
+	// isDestinationCaseInsensitive is true when the destination is case-insensitive
+	// In Windows, both paths D:\path\to\dir and D:\Path\TO\DiR point to the same resource.
+	// Apple File System (APFS) can be configured to be case-sensitive or case-insensitive.
+	// So for such locations, the key in the indexMap will be lowercase to avoid infinite syncing.
+	isDestinationCaseInsensitive bool
+}
+
+func newfolderIndexer() *folderIndexer {
+	return &folderIndexer{folderMap: make(map[string]*objectIndexer)}
+}
+
+// process the given stored object by indexing it using its relative path
+func (i *folderIndexer) store(storedObject StoredObject) (err error) {
+	// It is safe to index all StoredObjects just by relative path, regardless of their entity type, because
+	// no filesystem allows a file and a folder to have the exact same full path.  This is true of
+	// Linux file systems, Windows, Azure Files and ADLS Gen 2 (and logically should be true of all file systems).
+	var lcFileName, lcFolderName, lcRelativePath string
+
+	i.lock.Lock()
+	if i.isDestinationCaseInsensitive {
+		lcRelativePath = strings.ToLower(storedObject.relativePath)
+	} else {
+		lcRelativePath = storedObject.relativePath
+	}
+
+	if storedObject.isVirtualFolder || storedObject.entityType == common.EEntityType.Folder() {
+		lcFolderName = lcRelativePath
+	} else {
+		lcFolderName = filepath.Dir(lcRelativePath)
+		lcFileName = filepath.Base(lcRelativePath)
+	}
```

Comment:
We need to store root folder storedObject information. As it doesn't have parent, so we can't store storedObject. So either we make special case for root in that case it will be more confusing. I will add comment here to describe why it's done this way.

Reply:
Please ignore this comment. As i mentioned we store root folder as folderMap["."].indexMap["."] .